source/fe/mapping_isoparametric.cc

```diff
@@ -0,0 +1,2238 @@
+    dim,
+    spacedim>(computed_cell_similarity,
+              QProjector<dim>::DataSetDescriptor::cell(),
+              data,
+              output_data.jacobian_grads);
+
+  internal::MappingIsoparametricImplementation::
+    maybe_update_jacobian_pushed_forward_grads<dim, spacedim>(
+      computed_cell_similarity,
+      QProjector<dim>::DataSetDescriptor::cell(),
+      data,
+      output_data.jacobian_pushed_forward_grads);
+
+  internal::MappingIsoparametricImplementation::
+    maybe_update_jacobian_2nd_derivatives<dim, spacedim>(
+      computed_cell_similarity,
+      QProjector<dim>::DataSetDescriptor::cell(),
+      data,
+      output_data.jacobian_2nd_derivatives);
+
+  internal::MappingIsoparametricImplementation::
+    maybe_update_jacobian_pushed_forward_2nd_derivatives<dim, spacedim>(
+      computed_cell_similarity,
+      QProjector<dim>::DataSetDescriptor::cell(),
+      data,
+      output_data.jacobian_pushed_forward_2nd_derivatives);
+
+  internal::MappingIsoparametricImplementation::
+    maybe_update_jacobian_3rd_derivatives<dim, spacedim>(
+      computed_cell_similarity,
+      QProjector<dim>::DataSetDescriptor::cell(),
+      data,
+      output_data.jacobian_3rd_derivatives);
+
+  internal::MappingIsoparametricImplementation::
+    maybe_update_jacobian_pushed_forward_3rd_derivatives<dim, spacedim>(
+      computed_cell_similarity,
+      QProjector<dim>::DataSetDescriptor::cell(),
+      data,
+      output_data.jacobian_pushed_forward_3rd_derivatives);
+
+  const UpdateFlags          update_flags = data.update_each;
+  const std::vector<double> &weights      = quadrature.get_weights();
+
+  // Multiply quadrature weights by absolute value of Jacobian determinants or
+  // the area element g=sqrt(DX^t DX) in case of codim > 0
+
+  if (update_flags & (update_normal_vectors | update_JxW_values))
+    {
+      AssertDimension(output_data.JxW_values.size(), n_q_points);
+
+      Assert(!(update_flags & update_normal_vectors) ||
+               (output_data.normal_vectors.size() == n_q_points),
+             ExcDimensionMismatch(output_data.normal_vectors.size(),
+                                  n_q_points));
+
+
+      if (computed_cell_similarity != CellSimilarity::translation)
+        for (unsigned int point = 0; point < n_q_points; ++point)
+          {
+            if (dim == spacedim)
+              {
+                const double det = data.contravariant[point].determinant();
+
+#if false
+                // check for distorted cells.
+
+                // TODO: this allows for anisotropies of up to 1e6 in 3D and
+                // 1e12 in 2D. might want to find a finer
+                // (dimension-independent) criterion
+                // TODO[peterrum] : cell->diameter() not implemented yet
+                Assert(det >
+                         1e-12 * Utilities::fixed_power<dim>(
+                                   cell->diameter() / std::sqrt(double(dim))),
```

Comment:
The `TriaAccessor::diameter()` function still has to be implemented. However, since this is only an assert it could be postponed to a follow-up PR.